### ISSUE TITLE
DefaultBinding.pgFromString() doesn't support TIMESTAMP WITH TIME ZONE data type

### ISSUE THREAD

**Issue by sadanandk1** [jOOQ/jOOQ#6031]
i have created a simple converter and tried with a SP which returns table rows with timestamp with timezone column. it fails with number format exception as JOOQ code is trying to convert string with timezone info to java timestamp. 

```java
public class TimestampConverter implements Converter<Timestamp, Instant> {
    @Override public Instant from(Timestamp ts) {
        return ts == null ? null : ts.toInstant();
    }

    @Override public Timestamp to(Instant instant) {
        return instant == null ? null : Timestamp.from(null);
    }

    @Override public Class<Timestamp> fromType() { return Timestamp.class; }
    @Override public Class<Instant> toType() { return Instant.class; }
}
```

table:

```sql
 CREATE TABLE public.test
(
  timezone timestamp with time zone,
  notimezone timestamp without time zone,
  id numeric
)
```

SP:

```sql
CREATE OR REPLACE FUNCTION public.gettest(id_ bigint)
  RETURNS test AS
$BODY$

DECLARE
	test_ public.test%ROWTYPE;
BEGIN
	
	SELECT * INTO STRICT test_ FROM public.test WHERE id = id_;
	return test_;
END;
```

hand written Data class: 

```java
public class Test
{
	private Long Id;
	private Timestamp timezone;
	private Timestamp noTimezone;
	
	
	public Long getId() {
		return Id;
	}


	public void setId(Long id) {
		Id = id;
	}


	public Timestamp getTimezone() {
		return timezone;
	}


	public void setTimezone(Timestamp timezone) {
		this.timezone = timezone;
	}


	public Timestamp getNoTimezone() {
		return noTimezone;
	}


	public void setNoTimezone(Timestamp noTimezone) {
		this.noTimezone = noTimezone;
	}


	public Test()
	{
	}
	
}
```

POM.xml extract:

```xml
						<configuration>
							<jdbc>
								<driver>org.postgresql.Driver</driver>
								<url>jdbc:postgresql://localhost:5432/DBname</url>
								<user>test</user>
								<password>test</password>
							</jdbc>
							<generator>
								<name>org.jooq.util.DefaultGenerator</name>
								<database>
								    <customTypes>
					                    <customType>
					                        <name>Instant</name>
					                        <type>java.time.Instant</type>
					                        <converter>com.xxx.converter.TimestampConverter</converter>
					                    </customType>
					                </customTypes>
					
					                <forcedTypes>
					                    <forcedType>
					                        <name>Instant</name>
					                        <types>timestamp\ with\ time\ zone</types>
					                    </forcedType>
					                </forcedTypes>
									<name>org.jooq.util.postgres.PostgresDatabase</name>
									<includes>.*</includes>
									<excludes />
									  <schemata>
									    <schema>
									      <inputSchema>public</inputSchema>
									    </schema>
									  </schemata>
								</database>
								<target>
									<packageName>com.xxx.demo.jooq.domain</packageName>
									<directory>gensrc/main/java</directory>
								</target>
							</generator>
						</configuration>
```

now generate the code using jooq and call the above stored procedure as below
Routines.gettest(dsl.configuration(), Id).into(Test.class).

you will get this exception on timestamp with timezone column data.

```
  java.lang.NumberFormatException: For input string: "878472+00"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_111]
	at java.lang.Integer.parseInt(Integer.java:580) ~[na:1.8.0_111]
	at java.lang.Integer.parseInt(Integer.java:615) ~[na:1.8.0_111]
	at java.sql.Timestamp.valueOf(Timestamp.java:255) ~[na:1.8.0_111]
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2097) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2041) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.DefaultBinding.pgFromString(DefaultBinding.java:2130) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.DefaultBinding.pgSetValue(DefaultBinding.java:2269) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.DefaultBinding$2.operate(DefaultBinding.java:2162) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:128) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.DefaultBinding.pgNewRecord(DefaultBinding.java:2154) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.DefaultBinding.get(DefaultBinding.java:1479) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1640) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1611) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.operate(CursorImpl.java:1576) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:128) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchOne(CursorImpl.java:1540) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1507) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:329) ~[jooq-3.8.4.jar:na]
	at org.jooq.impl.CursorImpl.fetch(CursorImpl.java:282) ~[jooq-3.8.4.jar:na]
```

**Comment by lukaseder** [comment]
Thank you very much for your report. May I please ask you to show:

- How you used it? (The jOOQ code where it was used)
- What your query was?
- What your database table looks like?

**Comment by sadanandk1** [comment]
i used version 3.8.4, generated code. Used it in a application using generated code Routines.pGetXXX(dsl.configuration(), Id). this procedure retuns table rows and couple of columns are of type timestamp with timezone. i saw that in JOOQ code, it was trying to convert string to timestamp which has timezone info using Timestamp.valueOf(String) and failing as this function doesn't support string with timezone info.

**Comment by sadanandk1** [comment]
i updated issue with all the information now.

**Comment by lukaseder** [comment]
Thank you very much for the details. I think the main issue here is that jOOQ currently doesn't support functions that return a single table row type in PostgreSQL. I cannot seem to find the relevant issue, so I've created a new one: https://github.com/jOOQ/jOOQ/issues/6057

The other thing that might be problematic here is your usage of a converter rather than a binding for the `TIMESTAMP WITH TIME ZONE` type. We've just recently improved our support for this data type in the latest jOOQ versions 3.9.x (and still many bugs need fixing). Perhaps, it might be better to 1) upgrade to jOOQ 3.9, and 2) use the built-in support for `java.time.OffsetDateTime`, which will be more appropriate to convert to `Instant`. Alternatively, you can implement a data type `Binding` for `Instant` yourself.

**Comment by lukaseder** [comment]
Hmm, nope, I was wrong. #6057 already works. So the problem must be related to the deserialisation of the `TIMESTAMP WITH TIME ZONE` data type over such a nested table.

**Comment by lukaseder** [comment]
Indeed, the `TIMESTAMP WITH TIME ZONE` data type is currently not supported yet from this access pattern. Will fix this immediately.

Thanks again for reporting in such detail!

**Comment by lukaseder** [comment]
This is now fixed for jOOQ 3.10. Will merge this into jOOQ 3.9.2, which will hopefully ship this week.

**Comment by lukaseder** [comment]
(note, the correct type for the database type `<T>` in your converter will then be `OffsetDateTime`, not `Timestamp`)
